pkg/clusteragent/custommetrics/store.go

```diff
@@ -0,0 +1,137 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2017 Datadog, Inc.
+
+// +build kubeapiserver
+
+package custommetrics
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+// Store is an interface for persistent storage of custom and external metrics.
+type Store interface {
+	SetExternalMetric(ExternalMetricValue) error
+	DeleteExternalMetric(hpaNamespace, hpaName, metricName string) error
+	ListAllExternalMetrics() ([]ExternalMetricValue, error)
+	Update() error
+}
+
+// configMapStore provides persistent storage of custom and external metrics using a configmap.
+type configMapStore struct {
+	namespace string
+	name      string
+	client    corev1.CoreV1Interface
+	cm        *v1.ConfigMap
+}
+
+// NewConfigMapStore returns a new store backed by a configmap. The configmap will be created
+// in the specified namespace if it does not exist.
+func NewConfigMapStore(client kubernetes.Interface, ns, name string) (Store, error) {
+	cm, err := client.CoreV1().ConfigMaps(ns).Get(name, metav1.GetOptions{})
+	if err == nil {
+		log.Infof("Retrieved the configmap %s", name)
+		return &configMapStore{
+			namespace: ns,
+			name:      name,
+			client:    client.CoreV1(),
+			cm:        cm,
+		}, nil
+	}
+
+	if !errors.IsNotFound(err) {
+		log.Infof("Error while attempting to fetch the configmap %s: %s", name, err)
+		return nil, err
+	}
+
+	log.Infof("The configmap %s does not exist, trying to create it", name)
+	cm = &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: ns,
+		},
+	}
+	// FIXME: distinguish RBAC error
+	cm, err = client.CoreV1().ConfigMaps(ns).Create(cm)
+	if err != nil {
+		return nil, err
+	}
+	return &configMapStore{
+		namespace: ns,
+		name:      name,
+		client:    client.CoreV1(),
+		cm:        cm,
+	}, nil
+}
+
+// SetExternalMetric updates the associated external metric in the cached configmap.
+func (c *configMapStore) SetExternalMetric(em ExternalMetricValue) error {
+	if c.cm == nil {
+		return fmt.Errorf("configmap not initialized")
+	}
+	if c.cm.Data == nil {
+		// Don't panic "assignment to entry in nil map" at init
+		c.cm.Data = make(map[string]string)
+	}
+	key := fmt.Sprintf("external_metric.%s.%s-%s", em.HPANamespace, em.HPAName, em.MetricName)
+	toStore, _ := json.Marshal(em)
+	c.cm.Data[key] = string(toStore)
+	return nil
+}
+
+// DeleteExternalMetric deletes the associated external metric from the cached configmap.
+func (c *configMapStore) DeleteExternalMetric(hpaNamespace, hpaName, metricName string) error {
```

Comment:
Is the reason you decorrelated the `DeleteExternalMetric` and the actual Update of the configmap because it's easier to test ?
I think we need to remove the entry from the CM (or at list, log an error and set the metric as invalid) regardless of successfully removing it from the local store. I'm afraid that a local update failing (should the pipeline change in the future) might lead to a configmap that is not valid.

Reply:
I don't update the configmap in `DeleteExternalMetric` and `SetExternalMetric` so we can do multiple updates in one apiserver request. 

I will look into the issues you brought up with consistency.